spark3/src/main/java/org/apache/iceberg/actions/Spark3CreateAction.java

```diff
@@ -0,0 +1,172 @@
+
+    try {
+      this.sourceTable = (V1Table) ((TableCatalog) sourceCatalog).loadTable(sourceTableIdent);
+      this.sourceCatalogTable = sourceTable.v1Table();
+    } catch (NoSuchTableException e) {
+      throw new IllegalArgumentException(String.format("Cannot not find source table %s", sourceTableIdent), e);
+    } catch (ClassCastException e) {
+      throw new IllegalArgumentException(String.format("Cannot use non-v1 table %s as a source", sourceTableIdent), e);
+    }
+    validateSourceTable(sourceCatalogTable);
+
+    this.sourceTableLocation = CatalogUtils.URIToString(sourceCatalogTable.storage().locationUri().get());
+  }
+
+  @Override
+  public CreateAction withProperties(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  @Override
+  public CreateAction withProperty(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  protected SparkSession spark() {
+    return spark;
+  }
+
+  protected String sourceTableLocation() {
+    return sourceTableLocation;
+  }
+
+  protected CatalogTable v1SourceTable() {
+    return sourceCatalogTable;
+  }
+
+  protected CatalogPlugin sourceCatalog() {
+    return sourceCatalog;
+  }
+
+  protected Identifier sourceTableName() {
+    return sourceTableIdent;
+  }
+
+  protected Transform[] sourcePartitionSpec() {
+    return sourceTable.partitioning();
+  }
+
+  protected StagingTableCatalog destCatalog() {
+    return destCatalog;
+  }
+
+  protected Identifier destTableName() {
+    return destTableIdent;
+  }
+
+  protected Map<String, String> additionalProperties() {
+    return additionalProperties;
+  }
+
+  private static void validateSourceTable(CatalogTable sourceTable) {
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
+
+    if (!ALLOWED_SOURCES.contains(sourceTableProvider)) {
+      throw new IllegalArgumentException(
+          String.format("Cannot create an Iceberg table from source provider: %s", sourceTableProvider));
+    }
+
+    if (sourceTable.storage().locationUri().isEmpty()) {
+      throw new IllegalArgumentException("Cannot create an Iceberg table from a source without an explicit location");
+    }
+  }
+
+  protected static Map<String, String> tableLocationProperties(String tableLocation) {
+    return ImmutableMap.of(
+        TableProperties.WRITE_METADATA_LOCATION, tableLocation + "/" + ICEBERG_METADATA_FOLDER,
+        TableProperties.WRITE_NEW_DATA_LOCATION, tableLocation
+    );
+  }
+
+  protected static void assignDefaultTableNameMapping(Table table) {
+    NameMapping nameMapping = MappingUtil.create(table.schema());
+    String nameMappingJson = NameMappingParser.toJson(nameMapping);
+    table.updateProperties().set(TableProperties.DEFAULT_NAME_MAPPING, nameMappingJson).commit();
+  }
+
+  private static StagingTableCatalog checkDestinationCatalog(CatalogPlugin catalog) {
```

Comment:
Why static?

Reply:
I think it makes sense to keep the API for child implementations small. We can always refactor to expose it, but if we don't need to then private static makes sense to me.

Reply:
private sounds good to me but I am not sure why it needs to be static as it is used in non-static context.